// @arch-ui/pagination/dist/pagination.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var core = require('@emotion/core');
var react = require('react');
var icons = require('@arch-ui/icons');
var layout = require('@arch-ui/layout');
var loading = require('@arch-ui/loading');
var button = require('@arch-ui/button');
var theme = require('@arch-ui/theme');

const Page = props => {
  const {
    onClick,
    value,
    isSelected
  } = props;

  const handleClick = () => {
    if (onClick) {
      onClick(value);
    }
  };

  return core.jsx(button.Button, _extends({}, props, {
    onClick: handleClick,
    css: isSelected ? {
      backgroundColor: theme.colors.primary,
      color: 'white'
    } : {}
  }));
};

function ariaPageLabelFn(page) {
  return `Go to page ${page}`;
}

const PageChildren = ({
  page,
  isLoading,
  isSelected
}) => {
  const [shouldShowLoading, setShouldShowLoading] = react.useState(false);
  react.useEffect(() => {
    if (isLoading && isSelected) {
      const id = setTimeout(() => {
        setShouldShowLoading(true);
      }, 200);
      return () => {
        clearTimeout(id);
        setShouldShowLoading(false);
      };
    }
  }, [page, isLoading, isSelected]);
  return shouldShowLoading ? core.jsx("div", {
    css: {
      height: 19
    }
  }, core.jsx(loading.LoadingSpinner, null)) : core.jsx("span", null, page);
};

const Pagination = (_ref) => {
  let {
    ariaPageLabel = ariaPageLabelFn,
    currentPage = 1,
    limit = 5,
    pageSize,
    total,
    isLoading,
    onChange
  } = _ref,
      props = _objectWithoutProperties(_ref, ["ariaPageLabel", "currentPage", "limit", "pageSize", "total", "isLoading", "onChange"]);

  const [allPagesVisible, setAllPagesVisible] = react.useState(false);

  const toggleAllPages = () => {
    setAllPagesVisible(wereAllPagesVisible => !wereAllPagesVisible);
  };

  const renderPages = () => {
    if (total <= pageSize) return [];
    let pages = [];
    let totalPages = Math.ceil(total / pageSize);
    let minPage = 1;
    let maxPage = totalPages;
    const moreCharacter = core.jsx("span", null, "\u2026");

    if (limit && limit < totalPages) {
      let rightLimit = Math.floor(limit / 2);
      let leftLimit = rightLimit + limit % 2 - 1;
      minPage = currentPage - leftLimit;
      maxPage = currentPage + rightLimit;

      if (minPage < 1) {
        maxPage = limit;
        minPage = 1;
      }

      if (maxPage > totalPages) {
        minPage = totalPages - limit + 1;
        maxPage = totalPages;
      }
    }

    const handleChange = page => {
      if (onChange) {
        setAllPagesVisible(false);
        onChange(page, {
          pageSize,
          total,
          minPage,
          maxPage
        });
      }
    }; // go to first


    if (minPage > 1) {
      pages.push(core.jsx(Page, {
        "aria-label": ariaPageLabel(1),
        key: "page_start",
        onClick: handleChange,
        value: 1
      }, moreCharacter));
    } // loop over range


    for (let page = minPage; page <= maxPage; page++) {
      const isSelected = page === currentPage;
      pages.push(core.jsx(Page, {
        "aria-label": ariaPageLabel(page),
        "aria-current": isSelected ? 'page' : null,
        key: `page_${page}`,
        isSelected: isSelected,
        onClick: handleChange,
        value: page
      }, core.jsx(PageChildren, {
        isLoading: isLoading,
        page: page,
        isSelected: isSelected
      })));
    } // go to last


    if (maxPage < totalPages) {
      pages.push(core.jsx(Page, {
        "aria-label": ariaPageLabel(totalPages),
        key: "page_end",
        onClick: handleChange,
        value: totalPages
      }, moreCharacter));
    } // return pages;


    return [core.jsx(Page, {
      "aria-label": "Go to previous page",
      key: "page_prev",
      onClick: handleChange,
      value: currentPage - 1,
      isDisabled: currentPage === 1
    }, core.jsx(icons.ChevronLeftIcon, null)), allPagesVisible ? pages : core.jsx(Page, {
      "aria-label": "Click to show all pages",
      key: "page_dot",
      onClick: toggleAllPages,
      id: "ks-pagination-show-pages",
      value: 1 // needs value for flow...

    }, core.jsx(icons.ListOrderedIcon, null)), core.jsx(Page, {
      "aria-label": "Go to next page",
      key: "page_next",
      onClick: handleChange,
      value: currentPage + 1,
      isDisabled: currentPage === totalPages
    }, core.jsx(icons.ChevronRightIcon, null))];
  };

  return core.jsx(layout.FlexGroup, _extends({
    as: "nav",
    align: "center",
    "aria-label": "Pagination",
    isContiguous: true,
    isInline: true
  }, props), renderPages());
};

exports.Pagination = Pagination;
