// @arch-ui/controls/dist/controls.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _extends = _interopDefault(require('@babel/runtime/helpers/extends'));
var React = require('react');
var React__default = _interopDefault(React);
var reactRadios = require('react-radios');
var _defineProperty = _interopDefault(require('@babel/runtime/helpers/defineProperty'));
var _objectWithoutProperties = _interopDefault(require('@babel/runtime/helpers/objectWithoutProperties'));
var reactPseudoState = require('react-pseudo-state');
var styled = _interopDefault(require('@emotion/styled'));
var theme = require('@arch-ui/theme');
var input = require('@arch-ui/input');

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
const Wrapper = styled.div({
  display: 'flex',
  alignItems: 'center'
});
const Label = styled.label({
  alignItems: 'center',
  display: 'flex',
  lineHeight: 1
});
const Text = 'span';
const Icon = styled.div(({
  checked,
  isDisabled,
  isFocus,
  isActive,
  isHover
}) => {
  // background
  let bg = theme.colors.N10;

  if (isDisabled && checked) {
    bg = theme.colors.N30;
  } else if (isActive) {
    bg = checked ? theme.colors.B.D10 : theme.colors.N20;
  } else if ((isFocus || isHover) && !checked) {
    bg = theme.colors.N15;
  } else if (checked) {
    bg = theme.colors.B.base;
  } // fill


  let fill = 'white';

  if (isDisabled && checked) {
    fill = theme.colors.N70;
  } else if (!checked) {
    fill = 'transparent';
  } // stroke


  let innerStroke = isFocus ? theme.colors.B.L20 : bg;
  let innerStrokeWidth = 1;

  if (checked && !isDisabled) {
    innerStroke = isActive ? theme.colors.B.D20 : theme.colors.B.base;
  }

  let outerStroke = 'transparent';
  let outerStrokeWidth = 1;

  if (isFocus && !isActive) {
    outerStroke = theme.colors.B.A20;
    outerStrokeWidth = 5;
  }

  return {
    color: bg,
    fill,
    lineHeight: 0,
    cursor: isDisabled ? 'not-allowed' : null,
    // awkwardly apply the focus ring
    '& .outer-stroke': {
      transition: 'stroke 0.2s ease-in-out',
      stroke: outerStroke,
      strokeWidth: outerStrokeWidth
    },
    '& .inner-stroke': {
      stroke: innerStroke,
      strokeWidth: innerStrokeWidth
    }
  };
});
const defaultComponents = {
  Wrapper,
  Label,
  Text
};
const Control = (_ref) => {
  let {
    checked = false,
    children,
    components: propComponents = {},
    isDisabled = false,
    isRequired,
    name,
    onChange,
    icon: IconContent,
    tabIndex,
    type,
    value,
    id
  } = _ref,
      wrapperProps = _objectWithoutProperties(_ref, ["checked", "children", "components", "isDisabled", "isRequired", "name", "onChange", "icon", "tabIndex", "type", "value", "id"]);

  const components = React.useMemo(() => _objectSpread({}, defaultComponents, {}, propComponents), [propComponents]);
  return React__default.createElement(components.Wrapper, wrapperProps, React__default.createElement(reactPseudoState.PseudoState, null, ({
    onBlur,
    onFocus,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseEnter,
    onMouseLeave,
    onMouseUp,
    onTouchEnd,
    onTouchStart
  }, snapshot) => {
    const labelHandlers = {
      onMouseDown,
      onMouseUp,
      onMouseEnter,
      onMouseLeave,
      onTouchEnd,
      onTouchStart
    };
    const inputHandlers = {
      onBlur,
      onChange,
      onFocus,
      onKeyDown,
      onKeyUp
    };

    const iconProps = _objectSpread({}, snapshot, {
      checked,
      isDisabled
    });

    return React__default.createElement(components.Label, _extends({
      isChecked: checked,
      isDisabled: isDisabled
    }, labelHandlers), React__default.createElement(input.HiddenInput, _extends({}, inputHandlers, {
      checked: checked,
      disabled: isDisabled,
      name: name,
      required: isRequired,
      tabIndex: tabIndex,
      type: type,
      value: value,
      id: id
    })), React__default.createElement(Icon, iconProps, React__default.createElement(IconContent, null)), children ? React__default.createElement(components.Text, null, children) : null);
  }));
};

const CheckboxIcon = React.memo(props => React.createElement("svg", _extends({
  focusable: "false",
  height: "24",
  role: "presentation",
  viewBox: "0 0 24 24",
  width: "24"
}, props), React.createElement("g", {
  fillRule: "evenodd"
}, React.createElement("rect", {
  className: "outer-stroke",
  fill: "transparent",
  x: "6",
  y: "6",
  width: "12",
  height: "12",
  rx: "2"
}), React.createElement("rect", {
  className: "inner-stroke",
  fill: "currentColor",
  x: "6",
  y: "6",
  width: "12",
  height: "12",
  rx: "2"
}), React.createElement("path", {
  d: "M9.707 11.293a1 1 0 1 0-1.414 1.414l2 2a1 1 0 0 0 1.414 0l4-4a1 1 0 1 0-1.414-1.414L11 12.586l-1.293-1.293z",
  fill: "inherit"
}))));
const RadioIcon = React.memo(props => React.createElement("svg", _extends({
  focusable: "false",
  height: "24",
  role: "presentation",
  viewBox: "0 0 24 24",
  width: "24"
}, props), React.createElement("g", {
  fillRule: "evenodd"
}, React.createElement("circle", {
  className: "outer-stroke",
  fill: "transparent",
  cx: "12",
  cy: "12",
  r: "7"
}), React.createElement("circle", {
  className: "inner-stroke",
  fill: "currentColor",
  cx: "12",
  cy: "12",
  r: "7"
}), React.createElement("circle", {
  fill: "inherit",
  cx: "12",
  cy: "12",
  r: "2"
}))));
const CheckboxPrimitive = props => React.createElement(Control, _extends({
  icon: CheckboxIcon,
  type: "checkbox"
}, props));
const RadioPrimitive = props => React.createElement(Control, _extends({
  icon: RadioIcon,
  type: "radio"
}, props));

const Checkbox = props => React__default.createElement(reactRadios.Checkbox, _extends({
  component: CheckboxPrimitive
}, props));
const CheckboxGroup = props => React__default.createElement(reactRadios.CheckboxGroup, props);
const Radio = props => React__default.createElement(reactRadios.Radio, _extends({
  component: RadioPrimitive
}, props));
const RadioGroup = props => React__default.createElement(reactRadios.RadioGroup, props);

exports.Checkbox = Checkbox;
exports.CheckboxGroup = CheckboxGroup;
exports.CheckboxPrimitive = CheckboxPrimitive;
exports.Radio = Radio;
exports.RadioGroup = RadioGroup;
exports.RadioPrimitive = RadioPrimitive;
