// @keystonejs/fields/dist/Field-a10ff972.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var fields = require('@arch-ui/fields');

var React = require('react');

var React__default = _interopDefault(React);

var input = require('@arch-ui/input');

const DecimalField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors
}) => {
  const handleChange = event => {
    const value = event.target.value;
    onChange(value.replace(/[^0-9.,]+/g, ''));
  };

  const valueToString = value => {
    // Make the value a string to prevent loss of accuracy and precision.
    if (typeof value === 'string') {
      return value;
    } else if (typeof value === 'number') {
      return String(value);
    } else {
      // If it is neither string nor number then it must be empty.
      return '';
    }
  };

  const {
    // currency,
    // digits,
    symbol
  } = field.config;
  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(fields.FieldContainer, null, React__default.createElement(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), React__default.createElement(fields.FieldDescription, {
    text: field.adminDoc
  }), React__default.createElement(fields.FieldInput, null, symbol && React__default.createElement(fields.Currency, null, symbol), React__default.createElement(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "text",
    value: valueToString(value),
    onChange: handleChange,
    id: htmlID
  })));
};

exports.default = DecimalField;
