// @keystonejs/fields/dist/Field-5f120e65.esm.js
import { FieldContainer, FieldLabel, FieldDescription, FieldInput, Currency } from '@arch-ui/fields';
import React__default from 'react';
import { Input } from '@arch-ui/input';

const DecimalField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors
}) => {
  const handleChange = event => {
    const value = event.target.value;
    onChange(value.replace(/[^0-9.,]+/g, ''));
  };

  const valueToString = value => {
    // Make the value a string to prevent loss of accuracy and precision.
    if (typeof value === 'string') {
      return value;
    } else if (typeof value === 'number') {
      return String(value);
    } else {
      // If it is neither string nor number then it must be empty.
      return '';
    }
  };

  const {
    // currency,
    // digits,
    symbol
  } = field.config;
  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(FieldContainer, null, React__default.createElement(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), React__default.createElement(FieldDescription, {
    text: field.adminDoc
  }), React__default.createElement(FieldInput, null, symbol && React__default.createElement(Currency, null, symbol), React__default.createElement(Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "text",
    value: valueToString(value),
    onChange: handleChange,
    id: htmlID
  })));
};

export default DecimalField;
