// @keystonejs/fields/dist/Controller-5a535eb3.esm.js
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/classCallCheck';
import 'lodash.isequal';
import FieldController from '../Controller/dist/fields.esm.js';

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

class RelationshipController extends FieldController {
  constructor(config, ...args) {
    const {
      defaultValue = config.many ? [] : null
    } = config;
    super(_objectSpread(_objectSpread({}, config), {}, {
      defaultValue
    }), ...args);

    _defineProperty(this, "getQueryFragment", (path = this.path) => {
      return `
      ${path} {
        id
        _label_
      }
    `;
    });

    _defineProperty(this, "getFilterGraphQL", ({
      type,
      value
    }) => {
      if (type === 'contains') {
        return `${this.path}_some: {id: "${value}"}`;
      } else if (type === 'is') {
        return `${this.path}: {id: "${value}"}`;
      }
    });

    _defineProperty(this, "getFilterLabel", ({
      label
    }) => {
      return `${this.label} ${label.toLowerCase()}`;
    });

    _defineProperty(this, "formatFilter", ({
      label,
      value
    }) => {
      return `${this.getFilterLabel({
        label
      })}: "${value}"`;
    });

    _defineProperty(this, "serialize", data => {
      const {
        path
      } = this;
      const {
        many
      } = this.config;
      let value = data[path];

      if (many) {
        let ids = [];

        if (Array.isArray(value)) {
          ids = value.map(x => x.id);
        }

        return {
          disconnectAll: true,
          connect: ids.map(id => ({
            id
          }))
        };
      }

      if (!value) {
        return {
          disconnectAll: true
        };
      }

      return {
        connect: {
          id: value.id
        }
      };
    });

    _defineProperty(this, "getFilterTypes", () => {
      const {
        many
      } = this.config;

      if (many) {
        return [{
          type: 'contains',
          label: 'Contains',
          getInitialValue: () => null
        }];
      } else {
        return [{
          type: 'is',
          label: 'Is',
          getInitialValue: () => null
        }];
      }
    });
  }

  getRefList() {
    return this.getListByKey(this.config.ref);
  }

}

export default RelationshipController;
