// @keystonejs/fields/dist/Field-dc13593f.esm.js
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import _extends from '@babel/runtime/helpers/esm/extends';
import { borderRadius, colors, gridSize } from '@arch-ui/theme';
import { Input } from '@arch-ui/input';
import { P as Preview } from './preview-cef0e272.esm.js';

const StyledPreview = _ref => {
  let {
    preview,
    originalUrl,
    fieldPath
  } = _ref,
      props = _objectWithoutProperties(_ref, ["preview", "originalUrl", "fieldPath"]);

  return jsx(Preview, _extends({
    data: preview,
    originalUrl: originalUrl,
    fieldPath: fieldPath,
    css: {
      backgroundColor: 'white',
      borderRadius,
      border: `1px solid ${colors.N20}`,
      marginTop: gridSize,
      padding: 4,
      width: 410,
      // 300px image + chrome
      boxSizing: 'border-box'
    }
  }, props));
};

const PlaceholderPreview = ({
  originalUrl,
  fieldPath
}) => jsx(StyledPreview, {
  data: {
    html: '<div style="background-color: darkgray; height: 2.5em" />',
    title: 'Preview will be generated after save'
  },
  originalUrl: originalUrl,
  fieldPath: fieldPath,
  css: {
    opacity: 0.3
  }
});

const OEmbedField = ({
  onChange,
  autoFocus,
  field,
  value = null,
  savedValue = null,
  errors
}) => {
  const handleChange = event => {
    onChange({
      originalUrl: event.target.value
    });
  };

  const htmlID = `ks-oembed-${field.path}`;
  const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
  const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
  const hasChanged = field.hasChanged({
    [field.path]: savedValue
  }, {
    [field.path]: value
  });
  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), jsx(FieldDescription, {
    text: field.adminDoc
  }), jsx(FieldInput, null, jsx(Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "url",
    value: canRead && value && value.originalUrl || '',
    placeholder: canRead ? undefined : error.message,
    onChange: handleChange,
    id: htmlID
  })), value && value.originalUrl && hasChanged && jsx(PlaceholderPreview, {
    originalUrl: value.originalUrl,
    fieldPath: field.path
  }), value && value.originalUrl && !hasChanged && jsx(StyledPreview, {
    preview: value.preview,
    originalUrl: value.originalUrl,
    fieldPath: field.path
  }));
};

export default OEmbedField;
