// @keystonejs/fields/dist/Field-6c0e70c5.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var fields = require('@arch-ui/fields');

var React = require('react');

var React__default = _interopDefault(React);

var input = require('@arch-ui/input');

const FloatField = ({
  onChange,
  autoFocus,
  field,
  value,
  errors
}) => {
  const handleChange = event => {
    const value = event.target.value; // Similar implementation as per old Keystone version

    if (/^-?\d*\.?\d*$/.test(value)) {
      onChange(value);
    }
  };

  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(fields.FieldContainer, null, React__default.createElement(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), React__default.createElement(fields.FieldDescription, {
    text: field.adminDoc
  }), React__default.createElement(fields.FieldInput, null, React__default.createElement(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "text",
    value: value,
    onChange: handleChange,
    id: htmlID
  })));
};

exports.default = FloatField;
