// @keystonejs/fields/dist/Field-d217b273.cjs.js
'use strict';

var core = require('@emotion/core');

var fields = require('@arch-ui/fields');

var input = require('@arch-ui/input');
/** @jsx jsx */


const TextField = ({
  onChange,
  autoFocus,
  field,
  errors,
  value: serverValue
}) => {
  const handleChange = event => {
    onChange(event.target.value);
  };

  const {
    isMultiline
  } = field.config;
  const value = serverValue || '';
  const htmlID = `ks-input-${field.path}`;
  const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
  const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, null, core.jsx(input.Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "text",
    value: canRead ? value : undefined,
    placeholder: canRead ? undefined : error.message,
    onChange: handleChange,
    id: htmlID,
    isMultiline: isMultiline
  })));
};

exports.default = TextField;
