// @keystonejs/fields/dist/Field-a9abaa67.esm.js
import { jsx } from '@emotion/core';
import { FieldContainer, FieldLabel, FieldDescription, FieldInput } from '@arch-ui/fields';
import { Input } from '@arch-ui/input';
/** @jsx jsx */

const UnsplashField = ({
  onChange,
  autoFocus,
  field,
  errors,
  value: serverValue
}) => {
  const handleChange = event => {
    onChange(event.target.value);
  };

  const value = serverValue || '';
  const htmlID = `ks-unsplash-${field.path}`;
  const canRead = errors.every(error => !(error instanceof Error && error.name === 'AccessDeniedError'));
  const error = errors.find(error => error instanceof Error && error.name === 'AccessDeniedError');
  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), jsx(FieldDescription, {
    text: field.adminDoc
  }), jsx(FieldInput, null, jsx(Input, {
    autoComplete: "off",
    autoFocus: autoFocus,
    type: "text",
    value: canRead ? value : undefined,
    placeholder: canRead ? 'Unsplash Image ID' : error.message,
    onChange: handleChange,
    id: htmlID
  })));
};

export default UnsplashField;
