// @keystonejs/fields/dist/Filter-8a420ddc.esm.js
import '@babel/runtime/helpers/esm/defineProperty';
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@emotion/core';
import React__default from 'react';
import '@babel/runtime/helpers/esm/extends';
import '@arch-ui/select';
import '@apollo/react-hooks';
import 'graphql-tag';
import 'react-select';
import 'intersection-observer';
import { R as RelationshipSelect } from './RelationshipSelect-cd3e6aaf.esm.js';

const EventCatcher = ({
  children
}) => React__default.createElement("div", {
  onClick: e => {
    e.preventDefault();
    e.stopPropagation();
  }
}, children);

const RelationshipFilterView = ({
  onChange,
  filter,
  field,
  value
}) => {
  const handleChange = option => {
    if (option === null) {
      onChange(null);
    } else {
      const {
        value
      } = option;

      if (value) {
        onChange(value.id);
      }
    }
  };

  if (!filter) return null;
  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(EventCatcher, null, React__default.createElement(RelationshipSelect, {
    field: field,
    renderContext: null,
    htmlID: htmlID,
    onChange: handleChange,
    value: value,
    isMulti: false
  }));
};

export default RelationshipFilterView;
