// @keystonejs/fields/dist/Filter-5eea1b5a.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

require("@babel/runtime/helpers/defineProperty");

require("@babel/runtime/helpers/objectWithoutProperties");

require('@emotion/core');

var React = require('react');

var React__default = _interopDefault(React);

require("@babel/runtime/helpers/extends");

require('@arch-ui/select');

require('@apollo/react-hooks');

require('graphql-tag');

require('react-select');

require('intersection-observer');

var RelationshipSelect = require('./RelationshipSelect-34cff1e7.cjs.js');

const EventCatcher = ({
  children
}) => React__default.createElement("div", {
  onClick: e => {
    e.preventDefault();
    e.stopPropagation();
  }
}, children);

const RelationshipFilterView = ({
  onChange,
  filter,
  field,
  value
}) => {
  const handleChange = option => {
    if (option === null) {
      onChange(null);
    } else {
      const {
        value
      } = option;

      if (value) {
        onChange(value.id);
      }
    }
  };

  if (!filter) return null;
  const htmlID = `ks-input-${field.path}`;
  return React__default.createElement(EventCatcher, null, React__default.createElement(RelationshipSelect.RelationshipSelect, {
    field: field,
    renderContext: null,
    htmlID: htmlID,
    onChange: handleChange,
    value: value,
    isMulti: false
  }));
};

exports.default = RelationshipFilterView;
